core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorConcurrencyTest.scala

```diff
@@ -0,0 +1,281 @@
+ */
+
+package kafka.coordinator.group
+
+import java.util.concurrent.{ConcurrentHashMap, TimeUnit}
+
+import kafka.common.OffsetAndMetadata
+import kafka.coordinator.AbstractCoordinatorConcurrencyTest
+import kafka.coordinator.AbstractCoordinatorConcurrencyTest._
+import kafka.coordinator.group.GroupCoordinatorConcurrencyTest._
+import kafka.server.{ DelayedOperationPurgatory, KafkaConfig }
+import kafka.utils._
+import kafka.utils.timer.MockTimer
+
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.internals.Topic
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.record.{ MemoryRecords, RecordBatch }
+import org.apache.kafka.common.requests.ProduceResponse.PartitionResponse
+import org.apache.kafka.common.requests.{ JoinGroupRequest, OffsetCommitRequest, OffsetFetchResponse, TransactionResult }
+
+import org.easymock.EasyMock
+import org.junit.Assert._
+import org.junit.{ After, Assert, Before, Test }
+
+import scala.collection._
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, Future, Promise, TimeoutException }
+
+
+class GroupCoordinatorConcurrencyTest extends AbstractCoordinatorConcurrencyTest[GroupMember] {
+  val nGroups = nThreads * 5
+  val nMembersPerGroup = nThreads
+
+  var groupCoordinator: GroupCoordinator = null
+  var groupMembers: Seq[GroupMember] = null
+
+  @Before
+  override def setUp() {
+    super.setUp()
+
+    // make two partitions of the group topic to make sure some partitions are not owned by the coordinator
+    EasyMock.expect(zkUtils.getTopicPartitionCount(Topic.GROUP_METADATA_TOPIC_NAME))
+      .andReturn(Some(numPartitions))
+      .anyTimes()
+    EasyMock.replay(zkUtils)
+
+    serverProps.setProperty(KafkaConfig.GroupMinSessionTimeoutMsProp, ConsumerMinSessionTimeout.toString)
+    serverProps.setProperty(KafkaConfig.GroupMaxSessionTimeoutMsProp, ConsumerMaxSessionTimeout.toString)
+    serverProps.setProperty(KafkaConfig.GroupInitialRebalanceDelayMsProp, GroupInitialRebalanceDelay.toString)
+
+    val config = KafkaConfig.fromProps(serverProps)
+
+    val heartbeatPurgatory = new DelayedOperationPurgatory[DelayedHeartbeat]("Heartbeat", timer, config.brokerId, reaperEnabled = false)
+    val joinPurgatory = new DelayedOperationPurgatory[DelayedJoin]("Rebalance", timer, config.brokerId, reaperEnabled = false)
+
+    groupCoordinator = GroupCoordinator(config, zkUtils, replicaManager, heartbeatPurgatory, joinPurgatory, timer.time)
+    groupCoordinator.startup(false)
+
+    groupMembers = (0 until nGroups).flatMap { i =>
+      new Group("group" + i, groupCoordinator, replicaManager, timer, nMembersPerGroup).members
+    }
+  }
+
+  @After
+  override def tearDown() {
+    try {
+      if (groupCoordinator != null)
+        groupCoordinator.shutdown()
+    } finally {
+      super.tearDown()
+    }
+  }
+
+  @Test
+  def verifyGoodPathConcurrency() {
+    val operations: Seq[GroupOperation[ _, _]] = Seq(
```

Comment:
I like the approach. But I guess we don't have any tests with multiple members in the same group yet?

Reply:
The tests create multiple members in each group (5 groups with 5 members each at the moment, run concurrently across 5 threads).